Support Python 3.12

## Description
Python 3.12 has been released on 2023-10-02: https://www.python.org/downloads/release/python-3120/

Azure CLI needs to support Python 3.12 like how we supported 3.10 (#19857), 3.11 (#24494).

## TODOs
- [x] [Knack](https://github.com/microsoft/knack) supports Python 3.12
  - https://github.com/microsoft/knack/pull/279
- [x] [azdev](https://github.com/Azure/azure-cli-dev-tools) supports Python 3.12
  - https://github.com/Azure/azure-cli-dev-tools/pull/458
- [x] Dependencies are updated to the Python 3.12 compatible version
  - `pymsalruntime` (dependency of `msal[broker]`) https://github.com/Azure/azure-cli/pull/28556
  - https://github.com/Azure/azure-cli/pull/29466
  - https://github.com/Azure/azure-cli/pull/29486
- [ ] CI passes with Python 3.12
  - Remove `distutils` https://github.com/Azure/azure-cli/pull/28796
  - https://github.com/Azure/azure-cli/pull/27196
  - https://github.com/Azure/azure-cli/pull/29504
  - https://github.com/Azure/azure-cli/pull/29515
  - https://github.com/Azure/azure-cli/pull/29517
  - Remove `assertDictContainsSubset` and `assertNotEquals` #28194
    - https://github.com/Azure/azure-cli/pull/29120
    - https://github.com/Azure/azure-cli/pull/29121
  - https://github.com/Azure/azure-cli/issues/28829
  - https://github.com/Azure/azure-cli/pull/29373
  - https://github.com/Azure/azure-cli/pull/29465
  - https://github.com/Azure/azure-cli/pull/29887
  - https://github.com/Azure/azure-cli-extensions/pull/8010
  - https://github.com/Azure/azure-cli-extensions/pull/8116

Just to provide quick update re: latest release of azure-cli i.e. version 2.58.0 (5th March 2024)
This version of azure-cli still fails to install on Windows with the same pymsalruntime fatal error LNK1120 errors that were reported back in November 2023

C:\windows\system32>pip install azure-cli --upgrade
Collecting azure-cli
  Downloading azure_cli-2.58.0-py3-none-any.whl.metadata (8.4 kB)
...
Collecting pymsalruntime<0.14,>=0.13.2 (from msal[broker]==1.26.0->azure-cli-core==2.58.0->azure-cli)
  Using cached pymsalruntime-0.13.12.tar.gz (1.1 MB)
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
...
Building wheels for collected packages: pymsalruntime
  Building wheel for pymsalruntime (pyproject.toml) ... error
  error: subprocess-exited-with-error

  × Building wheel for pymsalruntime (pyproject.toml) did not run successfully.
  │ exit code: 1
  ╰─> [50 lines of output]
...
"-IC:\Program Files (x86)\Windows Kits\NETFXSDK\4.8\include\um" /Tcpymsalruntime/PyMsalRuntime.c /Fobuild\temp.win-amd64-cpython-312\Release\pymsalruntime/PyMsalRuntime.obj
      PyMsalRuntime.c
      pymsalruntime/PyMsalRuntime.c(2888): warning C4013: 'PyUnicode_FromUnicode' undefined; assuming extern returning int
      pymsalruntime/PyMsalRuntime.c(2888): warning C4047: '=': 'PyObject *' differs in levels of indirection from 'int'
      pymsalruntime/PyMsalRuntime.c(3230): warning C4013: 'PyUnicode_AsUnicode' undefined; assuming extern returning int
...
         Creating library build\temp.win-amd64-cpython-312\Release\pymsalruntime\pymsalruntime.cp312-win_amd64.lib and object build\temp.win-amd64-cpython-312\Release\pymsalruntime\pymsalruntime.cp312-win_amd64.exp
      PyMsalRuntime.obj : error LNK2001: unresolved external symbol PyUnicode_AsUnicode
      PyMsalRuntime.obj : error LNK2001: unresolved external symbol PyUnicode_FromUnicode
      build\lib.win-amd64-cpython-312\pymsalruntime\pymsalruntime.cp312-win_amd64.pyd : fatal error LNK1120: 2 unresolved externals
      error: command 'C:\\Program Files (x86)\\Microsoft Visual Studio\\2022\\BuildTools\\VC\\Tools\\MSVC\\14.37.32822\\bin\\HostX86\\x64\\link.exe' failed with exit code 1120

Attempting to perform a manual upgrade of dependency pymsalruntime... pip install pymsalruntime --upgrade and pip install pymsalruntime==0.14.0 also fails with the same error. I see the latest release (0.14.1) was yanked back in January 2024 https://pypi.org/project/pymsalruntime/0.14.1/

--- a/src/azure-cli-core/azure/cli/core/extension/__init__.py
+++ b/src/azure-cli-core/azure/cli/core/extension/__init__.py
@@ -8,7 +8,7 @@ import os
 import traceback
 import json
 import re
-from distutils.sysconfig import get_python_lib  # pylint: disable=deprecated-module
+from sysconfig import get_path
 
 import pkginfo
 from knack.config import CLIConfig
@@ -22,7 +22,7 @@ _CUSTOM_EXT_SYS_DIR = az_config.get('extension', 'sys_dir', None)
 EXTENSIONS_DIR = os.path.expanduser(_CUSTOM_EXT_DIR) if _CUSTOM_EXT_DIR else os.path.join(GLOBAL_CONFIG_DIR,
                                                                                           'cliextensions')
 DEV_EXTENSION_SOURCES = _DEV_EXTENSION_SOURCES.split(',') if _DEV_EXTENSION_SOURCES else []
-EXTENSIONS_SYS_DIR = os.path.expanduser(_CUSTOM_EXT_SYS_DIR) if _CUSTOM_EXT_SYS_DIR else os.path.join(get_python_lib(), 'azure-cli-extensions')
+EXTENSIONS_SYS_DIR = os.path.expanduser(_CUSTOM_EXT_SYS_DIR) if _CUSTOM_EXT_SYS_DIR else os.path.join(get_path("purelib"), 'azure-cli-extensions')
 
 EXTENSIONS_MOD_PREFIX = 'azext_'
 

--- a/src/azure-cli/azure/cli/command_modules/acs/_loadbalancer.py
+++ b/src/azure-cli/azure/cli/command_modules/acs/_loadbalancer.py
@@ -3,7 +3,7 @@
 # Licensed under the MIT License. See License.txt in the project root for license information.
 # --------------------------------------------------------------------------------------------
 
-from distutils.version import StrictVersion  # pylint: disable=deprecated-module
+from packaging.version import Version
 from types import SimpleNamespace
 from knack.log import get_logger
 
@@ -13,7 +13,7 @@ logger = get_logger(__name__)
 def set_load_balancer_sku(sku, kubernetes_version):
     if sku:
         return sku
-    if kubernetes_version and StrictVersion(kubernetes_version) < StrictVersion("1.13.0"):
+    if kubernetes_version and Version(kubernetes_version) < Version("1.13.0"):
         logger.warning('Setting load_balancer_sku to basic as it is not specified and kubernetes'
                        'version(%s) less than 1.13.0 only supports basic load balancer SKU\n',
                        kubernetes_version)

--- a/src/azure-cli/azure/cli/command_modules/acs/custom.py
+++ b/src/azure-cli/azure/cli/command_modules/acs/custom.py
@@ -25,7 +25,7 @@ import time
 import uuid
 import webbrowser
 import zipfile
-from distutils.version import StrictVersion  # pylint: disable=deprecated-module
+from packaging.version import Version
 from urllib.error import URLError
 from urllib.request import urlopen
 from azure.cli.command_modules.acs.maintenanceconfiguration import aks_maintenanceconfiguration_update_internal
@@ -159,7 +159,7 @@ def _aks_browse(
 
     return_msg = None
     # open portal view if addon is not enabled or k8s version >= 1.19.0
-    if StrictVersion(instance.kubernetes_version) >= StrictVersion('1.19.0') or (not addon_profile.enabled):
+    if Version(instance.kubernetes_version) >= Version('1.19.0') or (not addon_profile.enabled):
         subscription_id = get_subscription_id(cmd.cli_ctx)
         dashboardURL = (
             # Azure Portal URL (https://portal.azure.com for public cloud)

--- a/src/azure-cli/setup.py
+++ b/src/azure-cli/setup.py
@@ -12,9 +12,9 @@ import sys
 try:
     from azure_cli_bdist_wheel import cmdclass
 except ImportError:
-    from distutils import log as logger
+    import logging
 
-    logger.warn("Wheel is not available, disabling bdist_wheel hook")
+    logging.warning("Wheel is not available, disabling bdist_wheel hook")
     cmdclass = {}
 
 VERSION = "2.59.0"

